pkg/causality/conflict_detector.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package causality
+
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/pingcap/tiflow/engine/pkg/containers"
+	"github.com/pingcap/tiflow/pkg/causality/internal"
+)
+
+// ConflictDetector implements a logic that dispatches transaction
+// to different workers in a way that transactions modifying the same
+// keys are never executed concurrently and have their original orders
+// preserved.
+type ConflictDetector[Worker worker[Txn], Txn txnEvent] struct {
+	workers []Worker
+
+	// slots are used to find all unfinished transactions
+	// conflicting with an incoming transactions.
+	slots *internal.Slots[*internal.Node]
+
+	// finishedTxnQueue is the queue for all transactions that
+	// are already finished, but still waiting to be processed
+	// by the conflict detector, because they need to be processed
+	// so that transactions conflicting with them can be unblocked.
+	finishedTxnQueue *containers.SliceQueue[finishedTxn[Txn]]
+
+	// resolvedTxnQueue is the queue for all transactions that
+	// have been "resolved" (i.e., conflict free) but are not yet
+	// sent to the workers.
+	resolvedTxnQueue *containers.SliceQueue[resolvedTxn[Txn]]
+
+	// nextWorkerID is used to dispatch transactions round-robin.
+	nextWorkerID atomic.Int64
+
+	wg      sync.WaitGroup
+	closeCh chan struct{}
+}
+
+type finishedTxn[Txn txnEvent] struct {
```

Comment:
Is it possible to replace `finishedTxn and resolvedTxn ` with a generic version of `Node`,  which holds a `txn` field internally.